### ISSUE TITLE
Too many open files running Lantern

### ISSUE THREAD

**Issue by myleshorton** [getlantern/lantern#3987]
So this is interesting. While testing pre lazy loading on scroll version of the newsfeed UI, I got too many open files on the backend presumably because of all the connections initiated from Lantern on the backend. 

chained proxy at x.x.x.x:443 to tcp://da.feedsportal.com:80: Unable to dial server (trusted) chained proxy at x.x.x.x:443: dial tcp x.x.x.x:443: socket: too many open files on pass 0...continuing

**Comment by myleshorton** [comment]
FYI @oxtoacart and @fffw. This should be quite easy to reproduce with a page that just opens a ton of connections (like the early versions of the newsfeed hitting all those bookmarks).

**Comment by oxtoacart** [comment]
@myleshorton What's your `ulimit -n` ?

**Comment by myleshorton** [comment]
2560

On Wed, Apr 20, 2016 at 1:31 PM, oxtoacart <notifications@github.com> wrote:

> @myleshorton <https://github.com/myleshorton> What's your ulimit -n ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/lantern/issues/3987#issuecomment-212591817>
>



-- 
--
President
Brave New Software Project, Inc.
https://www.getlantern.org
pgp A998 2B6E EF1C 373E 723F A813 045D A255 901A FD89

**Comment by oxtoacart** [comment]
Whoa nelly, that's a reasonably high limit!

**Comment by oxtoacart** [comment]
I just saw something similar to this on mobile:

```
    LanternMainActivity  D  Refresh feed clicked
                    art  F  art/runtime/indirect_reference_table.cc:77] Check failed: table_mem_map_.get() != nullptr ashmem_create_region failed for 'indirect ref table': Too many open files
```

The really weird thing is that I'm in airplane mode and don't have a working connection.  So, this might be a leak with us failing to close a connection somewhere.

**Comment by oxtoacart** [comment]
I don't think a web page could cause this, as browsers only open a limited number of connections at a time.  I ran with a pre-lazy-loading version (f59191deb24fea7dc4e9fef02e901eb96378e89a) and didn't see an inordinate number of connections (maxed out around 73, and I had 10 other tabs open).

I suspect that we may just have a leak somewhere.  I tried fetching config frequently and disabling my network connection, but that didn't show a leak.

**Comment by oxtoacart** [comment]
I left this running overnight, and it doesn't look like I have a file descriptor leak on desktop anyway.

**Comment by myleshorton** [comment]
Hey @oxtoacart thanks for looking at this. I'm experiencing this *right now* on desktop on the 2.2.0 release branch. Also note users are seeing this in #3875

It doesn't appear to be related to the browser to me. I'm getting this:

```
Apr 29 17:06:23.614 - 13m36s ERROR flashlight: pac.go:183 Unable to unset lantern as system proxy: Unable to execute /Users/.../Library/Application Support/byteexec/pac-cmd: pipe: too many open files
```

I wonder if the newsfeed is causing us to hit a lot more pages off the bat that are resulting in more resets of the system proxy, and the system proxy code itself is leaking file descriptors?

**Comment by oxtoacart** [comment]
Can you run lsof for that process and send me the results?

-- 
Cheers,
Ox

PGP Key - https://pgp.mit.edu/pks/lookup?op=get&search=ox@getlantern.org

-----------------------------------------------------------------------------------------

"I love people who harness themselves, an ox to a heavy cart,
who pull like water buffalo, with massive patience,
who strain in the mud and the muck to move things forward,
who do what has to be done, again and again."

- Marge Piercy

**Comment by myleshorton** [comment]
I stupidly killed the process thinking I could surely reproduce it, but no luck so far. I'll run lsof as soon as I get it again though.

**Comment by fffw** [comment]
So far there're two causes I've seen: CLOSED connections to 8787 port, and CLOSE_WAIT connections to 443 port of proxy server. The later didn't happen for long. CLOSED connections to 8787 happens often, though not always trigger `too many open files`.

**Comment by fffw** [comment]
It becomes a blocker even for 2.2.0. When running the binary built from 2433dab for 5 hours, the error happens and Lantern stops working.

When I click on the systray, following signal triggers panic. I'm looking at the goroutines.

> HIToolbox is damaged with error 4294967254 and TTheme instance 5284510. Aborting...
> SIGABRT: abort
> PC=0x7fff90798f06 m=0
> signal arrived during cgo execution

**Comment by oxtoacart** [comment]
@fffw How quickly does this show up?  I just tried running the UI and even did a boom test, and I was never able to see any connections in the CLOSE state.  I tried with both Go 1.5.4 and GO 1.6.2.

```
boom -x http://localhost:8787 -disable-keepalive -c 100 -n 5000 -q 10 https://feeds.getiantem.org/en_US/feed.json
```

**Comment by oxtoacart** [comment]
Just to confirm, the CLOSED connection file descriptors are held by lantern, not the browser, right?

**Comment by fffw** [comment]
@oxtoacart It should only happen with keep alive. In handler.go, we only close connections after piped data from client to proxy server, but not vise versa. With keep alive, browser may reuse connection unless server actively closes it. It may be one of the cause of high number of connections in http_proxy_lantern!

The call to `closeFunc` is intentionally removed in [6f9e636](https://github.com/getlantern/lantern/commit/6f9e63687a6bf16ff7099f93fe4fae30d4dddc4f#diff-4a9300068d35a2ca92619d266ea30e2fR122), presumably to fix some other issue. I added it back and see only 1 CLOSE state connection to 8787 port after 2 hours.

**Comment by fffw** [comment]
`CLOSED` connections are seen from `lsof -nPp <lantern-pid>`, but interesting enough `netstat -anp tcp` doesn't have that entry. It's still unknown why `CLOSED` file descriptors are not removed.

**Comment by fffw** [comment]
Hmm my hypothesis is wrong. The sending half of connection is closed as defer statement at handler.go#84.

**Comment by oxtoacart** [comment]
This might help https://github.com/sockjs/sockjs-node/issues/99

-- 
Cheers,
Ox

PGP Key - https://pgp.mit.edu/pks/lookup?op=get&search=ox@getlantern.org

-----------------------------------------------------------------------------------------

"I love people who harness themselves, an ox to a heavy cart,
who pull like water buffalo, with massive patience,
who strain in the mud and the muck to move things forward,
who do what has to be done, again and again."

- Marge Piercy

**Comment by fffw** [comment]
Ok finally found the reason. `dualFetcher.Do` doesn't close response body when bad status code received. That explains why only @myleshorton and I encountered the error: Both of us frequently got 400 error when fetching config via chained server due to the UserId issue (fix is in devel but not in release-2.2.0).
